extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -14,43 +14,414 @@
+    });
+
+    return this.mutateElement(() => {
+      this.renderResults_();
+    });
+  }
+
+  /**
+   * Create and return <div> element from given plan-text item.
+   * @param {string} item
+   * @return {!Element}
+   * @private
+   */
+  createElementFromItem_(item) {
+    const element = this.element.ownerDocument.createElement('div');
+    element.classList.add('i-amphtml-autocomplete-item');
+    element.setAttribute('role', 'listitem');
+    element.textContent = item;
+    return element;
+  }
+
+  /**
+  * Handle rendering results on user input.
+  * @return {!Promise}
+  * @private
+  */
+  inputHandler_() {
+    return this.mutateElement(() => {
+      this.renderResults_();
+      this.toggleResults(true);
+    });
+  }
+
+  /**
+   * Handle selecting items on user mousedown.
+   * @param {Event} event
+   * @return {!Promise}
+   * @private
+   */
+  selectHandler_(event) {
+    if (!this.isItemElement(event.target)) {
+      return Promise.resolve();
+    }
+    return this.mutateElement(() => {
+      this.selectItem(event.target);
+    });
+  }
+
+  /**
+   * Render filtered results on the current input and update the container_.
+   * @private
+   */
+  renderResults_() {
+    const userInput = this.inputElement_.value;
+    this.clearAllItems();
+    if (userInput.length < this.minChars_ || !this.inlineData_) {
+      return;
+    }
+    const filteredData = this.filterData_(this.inlineData_, userInput);
+    filteredData.forEach(item => {
+      this.container_.appendChild(this.createElementFromItem_(item));
+    });
+  }
+
+  /**
+   * Apply the filter to the given data based on the given input.
+   * @param {!Array<string>} data
+   * @param {string} input
+   * @return {!Array<string>}
+   * @private
+   */
+  filterData_(data, input) {
+    input = input.toLowerCase();
+    let filteredData = data.filter(item => {
+      switch (this.filter_) {
+        case FilterType.SUBSTRING:
+          return includes(item, input);
+        case FilterType.PREFIX:
+          return startsWith(item, input);
+        case FilterType.TOKEN_PREFIX:
+          return item.split(' ').some(token => {
+            return startsWith(token, input);
+          });
+        case FilterType.FUZZY:
+          throw new Error(`Filter not yet supported: ${this.filter_}`);
+        case FilterType.CUSTOM:
+          throw new Error(`Filter not yet supported: ${this.filter_}`);
+        case FilterType.NONE:
+          // Query server endpoint.
+          throw new Error(`Filter not yet supported: ${this.filter_}`);
+        default:
+          throw new Error(`Unexpected filter: ${this.filter_}`);
+      }
+    });
+
+    // Truncate to max-entries.
+    if (this.maxEntries_ && this.maxEntries_ < filteredData.length) {
+      filteredData = filteredData.slice(0, this.maxEntries_);
+    }
+
+    return filteredData;
+  }
+
+  /**
+   * Shows or hides the results container_.
+   * @param {boolean=} opt_display
+   */
+  toggleResults(opt_display) {
```

Comment:
nit: is this method not `@private` so it can be called in tests? if that's the only reason, you can annotate it as `@private visible for testing`

Reply:
Some of these are only not `@private` because I inconsistently remembered to add that tag when adding new methods. In general is it better practice to make every method private by default?

Reply:
Generally yes, except for the inherited methods from `AMP.BaseElement`. There are some cases for public methods, like `amp-carousel` exposes `interactionNext` and `interactionPrev` so they can be used by `amp-lightbox-gallery`